exchange/exchange_test.go

```diff
@@ -4212,6 +4238,195 @@ func TestCallSignHeader(t *testing.T) {
+		me: metricsConf.NewMetricsEngine(&config.Configuration{}, openrtb_ext.CoreBidderNames(), nil, nil),
+	}
+	testCases := []struct {
+		description                 string
+		givenBid                    *entities.PbsOrtbBid
+		givenBidResponseExt         *openrtb_ext.ExtBidResponse
+		givenBidderName             string
+		givenPubID                  string
+		expectedBannerCreativeValid bool
+		expectedBidAdMValid         bool
+	}{
+		{
+			description:                 "The dimensions and adm of the bid are invalid, expect false",
+			givenBid:                    &entities.PbsOrtbBid{Bid: &openrtb2.Bid{W: 200, H: 200, AdM: "http://domain.com/invalid"}},
+			givenBidResponseExt:         &openrtb_ext.ExtBidResponse{Errors: make(map[openrtb_ext.BidderName][]openrtb_ext.ExtBidderMessage)},
+			givenBidderName:             "bidder",
+			givenPubID:                  "1",
+			expectedBannerCreativeValid: false,
+			expectedBidAdMValid:         false,
+		},
+		{
+			description:                 "The dimensions and adm of the bid are valid, expect true",
+			givenBid:                    &entities.PbsOrtbBid{Bid: &openrtb2.Bid{W: 50, H: 50, AdM: "https://domain.com/valid"}},
+			givenBidResponseExt:         &openrtb_ext.ExtBidResponse{Errors: make(map[openrtb_ext.BidderName][]openrtb_ext.ExtBidderMessage)},
+			givenBidderName:             "bidder",
+			givenPubID:                  "1",
+			expectedBannerCreativeValid: true,
+			expectedBidAdMValid:         true,
+		},
+		{
+			description:                 "Dimensions are valid. The adm is invalid. Expect true/false",
+			givenBid:                    &entities.PbsOrtbBid{Bid: &openrtb2.Bid{W: 50, H: 50, AdM: "http://domain.com/invalid"}},
+			givenBidResponseExt:         &openrtb_ext.ExtBidResponse{Errors: make(map[openrtb_ext.BidderName][]openrtb_ext.ExtBidderMessage)},
+			givenBidderName:             "bidder",
+			givenPubID:                  "1",
+			expectedBannerCreativeValid: true,
+			expectedBidAdMValid:         false,
+		},
+		{
+			description:                 "Dimensions are invalid. The adm is valid. Expect false/true",
+			givenBid:                    &entities.PbsOrtbBid{Bid: &openrtb2.Bid{W: 200, H: 200, AdM: "https://domain.com/valid"}},
+			givenBidResponseExt:         &openrtb_ext.ExtBidResponse{Errors: make(map[openrtb_ext.BidderName][]openrtb_ext.ExtBidderMessage)},
+			givenBidderName:             "bidder",
+			givenPubID:                  "1",
+			expectedBannerCreativeValid: false,
+			expectedBidAdMValid:         true,
+		},
+	}
+	for _, test := range testCases {
+		acutalBannerCreativeValid := exchange.validateBannerCreativeSize(test.givenBid, test.givenBidResponseExt, openrtb_ext.BidderName(test.givenBidderName), test.givenPubID, "enforce")
+		assert.Equal(t, test.expectedBannerCreativeValid, acutalBannerCreativeValid)
+
+		actualBidAdMValid := exchange.validateBidAdM(test.givenBid, test.givenBidResponseExt, openrtb_ext.BidderName(test.givenBidderName), test.givenPubID, "enforce")
```

Comment:
Can we separate `TestValidateBidForBidResponse` into two tests? One that asserts `validateBannerCreativeSize` and the other one `validateBidAdM` separately?

Reply:
Before approving I'd like to double check the expected values of the `validateBidAdM` unit tests: In issue #1288, @SyntaxNode  [commented](https://github.com/prebid/prebid-server/issues/1288#issuecomment-663616854) in regards of the `validateBidAdM` function's behavior something like: `"...For the other formats, we may want the validation to fail if there is any http:' or 'http%3A' (as proposed) but also that there must be a match on https:' or 'https%3A'. This will ensure urls without protocols defined are not considered secure..."`.

Does this mean we should expect every URL in the `"AdM"` field to be written at least once with a secure protocol? Or do we want at least one secure protocol URL in the entiriety of the `"AdM"` field? In other words, should we expect the test below to pass or fail?

**Pass (at least one secure protocol in the entiriety of the `AdM` field)**
```
  {
      desc: "AdM comes www.foo.com under 'http:' and www.bar.com under 'https:' because there's at least one URL with a secure protocol ('https://www.bar.com') test should pass",
      givenBid: &entities.PbsOrtbBid{
          Bid: &openrtb2.Bid{
              AdM: `<div id=\"container\"><a href=\"http:\/\/www.foo.com\/"><img src="some-source"\></a><a href=\"https:\/\/www.bar.com\/">buy now!</a></div>`,
          }
      },
      expectedToPass: true, //<-- pass validation
  },
```
**Or fail? (every URL to be written at least once with a secure protocol)**
```
  {
      desc: "AdM comes www.foo.com under 'http:' and www.bar.com under 'https:' expect validation to fail because of non-secure http://www.foo.com",
      givenBid: &entities.PbsOrtbBid{
          Bid: &openrtb2.Bid{
              AdM: `<div id=\"container\"><a href=\"http:\/\/www.foo.com\/"><img src="some-source"\></a><a href=\"https:\/\/www.bar.com\/">buy now!</a></div>`,
          }
      },
      expectedToPass: false, //<-- fail validation
  },
```
To my understanding of issue #1288, it is agreed upon that the following test should pass:
```
  {
      desc: "AdM comes with www.foo.com preceeded by  'http:' in one instance and by 'https:' in another instance. Expect validation to pass",
      givenBid: &entities.PbsOrtbBid{
          Bid: &openrtb2.Bid{
              AdM: `<div id=\"container\"><a href=\"http:\/\/www.foo.com\/"><img src="some-source"\></a><a href=\"https:\/\/www.foo.com\/">buy now!</a></div>`,
          }
      },
      expectedToPass: true,
  },
```
Thoughts?

Reply:
Addressed in a meeting but the validation works as follows: Validation fails if it contains 'http:' or 'http%3A' and does not contain either 'https:' or 'https%3A'

This is how the Java version handles this part, and the Go version will handle it the same way